### PR TITLE
repr and get methods for AffineMap, w/ precise exceptions

### PR DESCRIPTION
**Pull request by raamana**
PR from discussion in #1442 

Changes pass the tests in dipy.align/tests/test_imaffine.py

Still need to discuss
 1. whether we should remove affine.affine, and make it private
 2. the alternative: @property and @property.setter etc

### CODE REVIEW

**Comment by raamana** [comment]
Usage is illustrated below:

```python
from dipy.align.imaffine import AffineMap
import numpy as np

init = np.random.random([4, 4])
a = AffineMap(init)

a

array([[0.87399387, 0.36447063, 0.49235451, 0.24511999],
       [0.37959061, 0.33566857, 0.84241025, 0.49286616],
       [0.92397491, 0.67201834, 0.78806139, 0.95720898],
       [0.46861731, 0.12886398, 0.38019741, 0.83287818]])

print(a)

[[0.87399387 0.36447063 0.49235451 0.24511999]
 [0.37959061 0.33566857 0.84241025 0.49286616]
 [0.92397491 0.67201834 0.78806139 0.95720898]
 [0.46861731 0.12886398 0.38019741 0.83287818]]

print('Full:\n{a:full}\nRotation:\n{a:rotation}\nTranslation:\n{a:translation}'.format(a=a))

Full:
[[0.87399387 0.36447063 0.49235451 0.24511999]
 [0.37959061 0.33566857 0.84241025 0.49286616]
 [0.92397491 0.67201834 0.78806139 0.95720898]
 [0.46861731 0.12886398 0.38019741 0.83287818]]
Rotation:
[[0.87399387 0.36447063 0.49235451]
 [0.37959061 0.33566857 0.84241025]
 [0.92397491 0.67201834 0.78806139]]
Translation:
[[0.24511999]
 [0.49286616]
 [0.95720898]]

print('Full:\n{a:f}\nRotation:\n{a:r}\nTranslation:\n{a:t}'.format(a=a))

Full:
[[0.87399387 0.36447063 0.49235451 0.24511999]
 [0.37959061 0.33566857 0.84241025 0.49286616]
 [0.92397491 0.67201834 0.78806139 0.95720898]
 [0.46861731 0.12886398 0.38019741 0.83287818]]
Rotation:
[[0.87399387 0.36447063 0.49235451]
 [0.37959061 0.33566857 0.84241025]
 [0.92397491 0.67201834 0.78806139]]
Translation:
[[0.24511999]
 [0.49286616]
 [0.95720898]]
```

Setter throws errors invalid input or elements or non-invertible matrices:

```python

a.set_affine(1)

Traceback (most recent call last):
  File "/home/praamana/anaconda2/envs/py36/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-12-f67de536707e>", line 1, in <module>
    a.set_affine(1)
  File "/home/praamana/dev/dipy/dipy/align/imaffine.py", line 171, in set_affine
    raise TypeError('Input is not of type ndarray.')
TypeError: Input is not of type ndarray.


init[1,2] = np.nan
a.set_affine(init)

Traceback (most recent call last):
  File "/home/praamana/anaconda2/envs/py36/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-14-3e7f2be45bd5>", line 1, in <module>
    a.set_affine(init)
  File "/home/praamana/dev/dipy/dipy/align/imaffine.py", line 174, in set_affine
    raise AffineInvalidValuesError('Affine contains invalid elements')
dipy.align.imaffine.AffineInvalidValuesError: Affine contains invalid elements
```

File: dipy/align/imaffine.py [hunk 1]
```diff
@@ -144,17 +160,61 @@ def set_affine(self, affine):
...
+                #   i.e. rows from 0 to n-1, cols from n to n
+                return str(self.affine[:-1,-1:])
+            else:
+                allowed_formats_print_map = ['full', 'f',
+                                             'rotation', 'r',
+                                             'translation', 't']
+                raise NotImplementedError('Format {} not recognized or implemented.\n'
+                                          'Try one of {}'.format(format_spec, allowed_formats_print_map))
+
+
```
Comment (skoudoro) [inline dipy/align/imaffine.py]: pep8: too many blank lines

**Comment by skoudoro** [comment]
Can you add some tests for  `__str__`, `__repr__`, and `__format__` ?

File: dipy/align/imaffine.py (hunk 1 above)
Comment (raamana) [reply dipy/align/imaffine.py]: isn't two lines between two methods standard? Atleast my Pycharm thinks so

Comment (skoudoro) [reply dipy/align/imaffine.py]: The best answer is here: https://www.python.org/dev/peps/pep-0008/#blank-lines

> Method definitions inside a class are surrounded by a single blank line.

**Comment by skoudoro** [comment]
Thank you for adding the tests! As you can see [here](https://codecov.io/gh/nipy/dipy/compare/6a78d508c74dd91271d632f3e6f2c7198c7f85ad...3b78830eb47e622047e41922dc94ae62a5174752/diff#D1-170) and [here](https://codecov.io/gh/nipy/dipy/compare/6a78d508c74dd91271d632f3e6f2c7198c7f85ad...3b78830eb47e622047e41922dc94ae62a5174752/diff#D1-230), you can improve them to make Codecov happy :-)

**Comment by raamana** [comment]
My bad, I didn't think about the test coverage :)

**Comment by skoudoro** [comment]
Great! Thanks @raamana! 

We still need to make a decision:
> whether we should remove affine.affine, and make it private
> the alternative: @property and @property.setter etc

Any opinion @arokem @matthew-brett @Garyfallidis @omarocegueda?

**Comment by raamana** [comment]
You’re welcome.

Perhaps, we can merge this for now, and plan a update/deprecation cycle for modifying attribute behaviour, however it’s voted to be? This PR is backwards compatible, and wouldn’t require changes in users code.

**Comment by skoudoro** [comment]
Ok, I will merge this PR in 3days if there is no objection.

**Comment by skoudoro** [comment]
Thanks @raamana! merging!

**Pull request merged**